File: client/lib/api/models/audio.dart
import 'dart:convert';

class Audio {
  String id;
  String userId;
  String contentType;
  String title;
  String fileName;
  String coverArtFileName;
  String bucketKey;
  String coverArtKey;
  int createdAt;
  int updatedAt;

  Audio({required this.id, required this.contentType, required this.title, required this.createdAt, required this.updatedAt, required this.userId, required this.fileName, required this.coverArtFileName, required this.bucketKey, required this.coverArtKey});

  factory Audio.fromJson(Map<String, dynamic> json) {
    final id = json['_id'];
    final contentType = json['contentType'];
    final title = json['title'];
    final userId = json['userId'];
    final fileName = json['fileName'];
    final coverArtFileName = json['coverArtFileName'];
    final bucketKey = json['bucketKey'];
    final coverArtKey = json['coverArtKey'];
    final createdAt = (json['createdAt'] as num).toInt();
    final updatedAt = (json['updatedAt'] as num).toInt();

    return Audio(id: id, contentType: contentType, title: title, createdAt: createdAt, updatedAt: updatedAt, userId: userId, fileName: fileName, coverArtFileName: coverArtFileName, bucketKey: bucketKey, coverArtKey: coverArtKey);
  }

  Map<String, dynamic> toJson() => ({'id': id, 'contentType': contentType, 'title': title, 'createdAt': createdAt, 'updatedAt': updatedAt, 'userId': userId, 'fileName': fileName, 'coverArtFileName': coverArtFileName, 'bucketKey': bucketKey, 'coverArtKey': coverArtKey});

  @override
  String toString() => jsonEncode(toJson());
}

File: backend/src/DB/models/Audio.ts
import { boolean, InferType, number, object, string } from 'yup';
import { contentTypeSchema, likeObjectId } from '../common_schemas';

export const collectionName = 'audio'

export const schemaVersion = 'v1.0.0'

const post = {
    title: string().required().label('Title'),
}
export const audioPostSchema = object().shape(post).required()

const create = {
    userId: likeObjectId.required(),
    contentType: contentTypeSchema.optional(),
    title: string().required().label('Title'),
    coverArtFileName: string().optional().label('Cover art file name'),
    bucketKey: string().optional().url(), // audio/<userId>/<audioId>
    webBucketKey: string().optional().url(), // audio/<userId>/web/<audioId>
    coverArtKey: string().optional().url(), // audio/cover_art/<userId>/<audioId>
    temporary: boolean().required(),
}
export const audioCreateSchema = object().shape(create).required()

const patch = {
    title: string().required().label('Title'),
}
export const audioPatchSchema = object().shape(patch).required()

const update = {
    schemaVersion: string().optional().min(6).max(20),
    userId: likeObjectId.optional(),

    contentType: contentTypeSchema.optional(),
    title: string().optional().label('Title'),
    coverArtFileName: string().optional().label('Cover art file name'),

    bucketKey: string().optional().url(), // audio/<userId>/<audioId>
    webBucketKey: string().optional().url(), // audio/<userId>/web/<audioId>
    coverArtKey: string().optional().url(), // audio/cover_art/<userId>/<audioId>

    temporary: boolean().optional(),
}
export const audioUpdateSchema = object().shape(update).required()

export const audioSchema = object().shape({
    schemaVersion: string().optional().min(6).max(20),
    _id: likeObjectId.optional(),

    userId: likeObjectId.required(),

    contentType: contentTypeSchema.optional(),
    title: string().required().label('Title'),
    coverArtFileName: string().optional().label('Cover art file name'),

    bucketKey: string().optional().url(), // audio/<userId>/<audioId>
    webBucketKey: string().optional().url(), // audio/<userId>/web/<audioId>
    coverArtKey: string().optional().url(), // audio/cover_art/<userId>/<audioId>

    temporary: boolean().required(),

    createdAt: number().optional(),
    updatedAt: number().optional(),
})

export type AudioPost = InferType<typeof audioPostSchema>
export type AudioCreate = InferType<typeof audioCreateSchema>
export type AudioPatch = InferType<typeof audioPatchSchema>
export type AudioUpdate = InferType<typeof audioUpdateSchema>
export type Audio = InferType<typeof audioSchema>
